pkg/cfn/builder/vpc_endpoint.go

```diff
@@ -166,10 +166,16 @@ func buildVPCEndpointServices(ec2API ec2iface.EC2API, region string, endpoints [
 		})
 
 		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "InvalidServiceName" {
+				return nil, &api.UnsupportedFeatureError{
```

Comment:
What's the advantage over `errors.Wrapf`?

Reply:
It's for the caller to identify errors of this type without relying on string comparison.

Reply:
In this case we only print the error anyway, right? I.e. why do we need the `errors.As` check?

Reply:
Not that it's an issue necessarily but is the handling of the error correct here?

Reply:
>In this case we only print the error anyway, right? I.e. why do we need the errors.As check?

We could just log the error here instead of wrapping it in a typed error, but I have avoided doing any I/O here because this code has no external side effects (if the provider implementation is mocked). We don't want logs for this case appearing in our unit tests. Additionally, the error type provides a standard format for unsupported features that the calling code can identify with a single check. The `errors.As` check is required for the caller to identify errors of this type because this check happens higher up in the chain where this error may already be wrapped by other errors.

Reply:
i don't mean logging it here but if you were to just return `errors.Wrap(err, fmt.Sprintf("fully-private clusters are not supported in region %q, please retry with a different region", region))` here or even `errors.New(fmt.Sprintf("fully-private clusters are not supported in region %q, please retry with a different region", region)`, wouldn't the user see the same thing they're seeing with this PR? We call `logger.Critical` in `pkg/ctl/create/cluster.go` anyway.
Like:
```
				ufe := &api.UnsupportedFeatureError{}
				if errors.As(err, &ufe) {
					logger.Critical(ufe.Message)
				}
				logger.Critical("%s\n", err.Error())
```
results in `logger.Critical` being called for this `UnsupportedFeatureError` twice basically

Reply:
With `errors.Wrap` or `errors.New`, the log message will not start with the "fully-private clusters are not supported" error,  the messages from the `errors.Wrap` call higher up in the chain will appear first, which can hide the user-friendly error message.

Having an error type with a `Message` field to represent this condition lets us log a more-specific, user-friendly message while still preserving the error context and logging it later.

Reply:
Ah the `UnsupportedFeatureError` itself is being wrapped again and isn't the highest level error? Now I get it.

Reply:
Yes, it isn't the highest-level error.